### DISCUSSION TITLE
combined return and if statement like ruby

### DISCUSSION THREAD

**Discussion by gregveres** [dotnet/csharplang#2195] [Language Ideas]
I was coming here to suggest a return <value> if <condition> type statement and Github suggested a [previous proposal from last summer](https://github.com/dotnet/csharplang/issues/1397). 

I don't think the original poster on that proposal articulated the right solution, but one of the [responses pointed out that Ruby has this construct already.](https://github.com/dotnet/csharplang/issues/1397#issuecomment-374693186).

I guess I should do some research into how well this construct performs in Ruby, but I can say that my code is littered with things like this:

```
var value = somefunction();
if (value == null) {
    return <error>;
}
```

I believe that the code would be more readable if that code could look like this:

```
var value = somefunction();
return <error> if (value == null);

```

Thanks

**Comment by svick** [comment]
I think https://github.com/dotnet/csharplang/issues/1785 is very close to what you're proposing (but there are quite a few proposals for similar features, so I'm not surprised it's not the one you found).

I think the main issues with this proposal are:

1. `if (value == null) return <error>;` is already valid. If it's a company code style, or something like that, that doesn't allow you to write it on one line, then you should consider changing the code style, instead of changing the whole language.
2. It has a fairly high chance of being confusing. Consider a line like:

    ```c#
    return "a string literal, or another expression whose character count is fairly considerable" if (value == null);
    ```

    If the `if` is at the end of the line, I think it effectively hides the most important information about the line. And that's a problem, especially when you're scanning code, when trying to understand what it does.

**Comment by Austin-bryan** [comment]
Instead of 

```csharp
var value = somefunction();
if (value == null) {
    return <error>;
}
```

Just do

```csharp
var value = somefunction();
if (value == null) return <error>;
```

Tada.

**Reply by AndrejZigmund** [reply]
Sure, but that obscures return point, which should be clearly visible by our coding standards. So our rules force that return/break/continue would be always first word on a line. And as we use Microsoft's Allman braces style we are forced to deal with big early returns.

I think the problems preventing it are quite synthetic - the syntactic sugar is successfully used in some other languages.

**Reply by CyrusNajmabadi** [reply]
> So our rules force that return/break/continue would be always first word on a line

That's a stylistic choice on your part.  We don't design with that in mind because it would mean every feature would need a 'next version' because someone rejects what is currently there for stylistic reasons. 

We do not mandate style. So your choices that limit existing applicable solutions are your own, and are on you to accept the outcomes. :-)

**Reply by AndrejZigmund** [reply]
This feature is just syntactic sugar, that allows you to be flexible in expressing your intent shortly. 
Say, extension methods. You can surely do without them - just use static methods. Java still "does well" without them, but I always see a lot of envy from ordinary Java programmers about the feature, and lots of stupid offenses on it from "Java purists", like "but now you can't distinguish a method", "but now you can can extend sealed classes" etc. In reality extension methods sugar is so useful and successful in real life cases, that c#14 rolls out MORE extension related features.

I see, that quite a lot of people wanted this thing, and I've seen clear and concise code in other languages that went like this
```
int SomeMethod(params)
{
    throw Ex1 if (cond0);
    return 0 if (cond1);
    return 1 if (cond2);
    ...
    return expr1 if (cond3);
    return ....(expr2);
}
void SomeVoid()
{
     return if (cond1);
     ....
}
```
One can surely do without this, one can surely workaround this, and one can surely go by bus, instead of driving a luxury car, but...

Then there're things like "catch... when" - and no one complains. Nobody tries to thinks up synthetic arguments "why it's bad", just because "another language was first to use it".

**Reply by CyrusNajmabadi** [reply]
> This feature is just syntactic sugar

It's not sugar here as it's really the same as the existing form, which you are choosing explicitly to not use. 

As I mentioned: 

> We don't design with that in mind because it would mean every feature would need a 'next version' because someone rejects what is currently there for stylistic reasons.

**Reply by AndrejZigmund** [reply]
> It's not sugar here as it's really the same as the existing form...

Well, why extension methods are considered syntactic sugar, and "return if" not? Aren't extension methods just the same as an existing form of a static method?
For me it's still the same point, which "Java purists" are using against extension methods.

**Reply by CyrusNajmabadi** [reply]
We didn't add extension methods because people were stylistically opposed to static method calls. 

> which "Java purists" are using against extension methods.

I have no interest in what java purists are our are not for. :-)

I'm simply explaining why the justification that this is needed because some people will not use the existing form due to stylistic preferences is not something we put stock into.

**Reply by AndrejZigmund** [reply]
> We didn't add extension methods because people were stylistically opposed to static method calls.

I wonder why extension methods were added then? Was it some feature, that "you couldn't write certain code without" or was it "for convenience of developers" (which IMHO should be core reason of some modern language features)

**Reply by CyrusNajmabadi** [reply]
> I wonder why extension methods were added then?

It was so that linq could be a pure syntactic transform.  if you have a `from x in y where x >= 0` we want to translate that *syntactically* to `y.Where(x => x >= 0)`.  As collections didn't have a .Where method on them, we needed a way to bridge that instance syntax to the actual static method declaration elsewhere.

We did *not* want linq to rewrite to be static calls because the entire point was that .Where would dispatch on the instance *if possible* (for example, for query providers), and that lambdas could become expression trees.

The point of Linq was to introduce *syntactic* patterns, that could still have unique domain semantics.  Extension methods were necessary for that to complete the story so that vanilla collections (all the way down to `IEnumerable<T>`) would light up using the exact same system without the linq transform having to know anything about this part. 

> Was it some feature, that "you couldn't write certain code without"

yes.  It was exactly that.  We could not do Linq in a proper fashion without extensions.

**Reply by CyrusNajmabadi** [reply]
> or was it "for convenience of developers" 

Again, we have something convenient for developers.  Choosing not to use it for stylistic reasons is not something we design around.  After all, say we give you the feature you want.  Then another person comes and says "i don't like either of those options.  give me `expr return if condition;`.  And now we have to go spend all the effort again because of their stylistic preferences.

**Reply by AndrejZigmund** [reply]
OK, that's very interesting. Thanks for discussion.

Guess nothing left, than just deal with it and "use the old ways".

**Reply by CyrusNajmabadi** [reply]
If tehre are other arguments in favor, then that's fine.  I'm just calling  out that style isn't part of the decision making process.  So it's good to leave those justifications aside :)

**Comment by spydacarnage** [comment]
All the examples I'm seeing in favour of this are the form
```cs
if (something is null) return <error>
```
which already has a succinct short form.

The one scenario where I've previously found myself wishing for a short-hand is the reverse, where I have a string of "validate" checks at the start of a method and if any of them return an error, return that. That does now have a shortening thanks to pattern matching, but it's still needing extra variable declarations that aren't otherwise needed. For example:
```cs
if (checkOne() is Error errorOne) return errorOne;
if (checkTwo() is Error errorTwo) return errorTwo;
if (checkThree() is Error errorThree) return errorThree;
if (checkFour() is Error errorFour) return errorFour;
```
vs.
```cs
return checkOne() if not null;  // not suggested syntax, just for intention
return checkTwo() if not null;
return checkThree() if not null;
return checkFour() if not null;
```

**Reply by CyrusNajmabadi** [reply]
we have syntax for that:

```c#
var error = checkOne() ?? checkTwo() ?? checkThree() ?? checkFour();
if (error != null)
   return error;
```

**Reply by spydacarnage** [reply]
It's not perfect, as I sometimes need to do other work in between the calls, but as with many situations, it meets the majority case, and that's good enough.
